Implement bytes

Martmists-GH opened the pull request:
- Added PyBytesType (based on PyStringType implementation)
- Added PyBytes
- Added IndexError for `PyBytes.__getitem__`

Martmists-GH committed:
implement bytes

ilevn commented on src/main/kotlin/interpreter/pyobject/types/PyBytesType.kt:
```diff
@@ -0,0 +1,45 @@
+object PyBytesType : PyType("bytes") {
+    override fun newInstance(kwargs: Map<String, PyObject>): PyObject {
+        val value = kwargs["value"] ?: error("Built-in signature mismatch")
+        when (value) {
```
Don't use a when for a single conditional check. Instead, use a single if or a takeIf/Unless closure

Martmists-GH replied:
more checks will come when the other possible parameter types are added

ilevn commented on src/main/kotlin/interpreter/pyobject/types/PyBytesType.kt:
```diff
@@ -0,0 +1,45 @@
+
+    /** bytes.__getitem__ */
+    val pyBytesIndex = PyBuiltinFunction.wrap("__getitem__", PyCallableSignature("index" to ArgType.POSITIONAL)) {
+        val self = it["self"]!!.cast<PyBytes>()
```
These should error out if the non-null assertion fails

Martmists-GH replied:
This is copied from `PyStringType`, and it can be assumed that this is always a valid cast.

Fuyukai replied:
The non-null assertion is fine because it'll never error out.

Fuyukai commented on src/main/kotlin/interpreter/pyobject/PyBytes.kt:
```diff
@@ -0,0 +1,22 @@
+
+import green.sailor.kython.interpreter.pyobject.types.PyBytesType
+
+class PyBytes(val wrapped: ByteArray) : PyObject(PyBytesType) {
```
Docstring.

ilevn replied:
This is also missing an `override`

Fuyukai commented on src/main/kotlin/interpreter/pyobject/PyBytes.kt:
```diff
@@ -0,0 +1,22 @@
+    override fun pyStr(): PyString = pyRepr()
+
+    override fun pyRepr(): PyString {
+        val inner = PyString("...")
```
This is not a valid bytes repr.

Fuyukai commented on src/main/kotlin/interpreter/pyobject/types/PyBytesType.kt:
```diff
@@ -0,0 +1,45 @@
+        try {
+            PyInt(self.wrapped[index.wrappedInt.toInt()].toLong())
+        } catch (e: IndexOutOfBoundsException) {
+            Exceptions.INDEX_ERROR("index out of range")
```
This doesn't do anything.

Martmists-GH replied:
this statement returns in the .wrap call, being either a PyInt or IndexError

Fuyukai replied:
You don't return it, you throw it

martmists replied:
oh, I just copied it from PyStringType.pyStrInt, will fix both

Fuyukai commented:
Once #4 is merged this also should have a wrapper type for the kyc object

Martmists-GH commented:
Requested changes have been made.

Note: Once #4 merges, the only change required (as far as I can tell) is to replace PyString with PyBytes on this line: https://github.com/Fuyukai/Kython/pull/4/files#diff-bc200d594526fd32c75b8ca445c87d44R115

ilevn commented on src/main/kotlin/interpreter/pyobject/PyBytes.kt:
```diff
@@ -0,0 +1,25 @@
+/**
+ * Represents a Python bytes object. This wraps a regular JVM ByteArray.
+ */
+class PyBytes(val wrapped: ByteArray) : PyObject(PyBytesType) {
```
Missing `override`

Fuyukai committed:
Add spotless linter.

Signed-off-by: Laura F Dickinson <l@veriny.tf>

Apply ktlint.

Signed-off-by: Laura F Dickinson <l@veriny.tf>

Add line length, fix some violations, make CircleCI run spotlessCheck.

Signed-off-by: Laura F Dickinson <l@veriny.tf>

Add spotless linter.

Signed-off-by: Laura F Dickinson <l@veriny.tf>

Apply ktlint.

Signed-off-by: Laura F Dickinson <l@veriny.tf>

Add line length, fix some violations, make CircleCI run spotlessCheck.

Signed-off-by: Laura F Dickinson <l@veriny.tf>

Add IndexError to map

Fix Bytes repr

docstring for PyBytes

throw exception

simplify repr

Make compatible with #4

Proper bytes repr

Make requested changes

Lint files

ilevn commented on src/main/kotlin/interpreter/pyobject/PyBytes.kt:
```diff
@@ -6,7 +6,7 @@ class PyBytes(val wrapped: ByteArray) : PyObject(PyBytesType) {
 
     override fun pyRepr(): PyString {
-        val inner = PyString("...")
+        val inner = PyString(wrapped.joinToString("") { it.toChar().toString() })
```
This could simply be `PyString(String(wrapped))` if I'm not mistaken.

Fuyukai replied:
No, both statements are wrong. repr() of a bytes encodes unprintable bytes to their hex escape. 

```
>>> b"\x01\x07\x7A"
b'\x01\x07z'
```

PyRepr() is also wrong (`"b'\\x01\\x07z'"`).

ilevn commented:
@martmists Sorry about that. I just saw your impl and didn't really take python's repr into consideration.

Fuyukai commented:
Please squash your commits down

Fuyukai commented on src/main/kotlin/interpreter/pyobject/PyBytes.kt:
```diff
@@ -0,0 +1,25 @@
+package green.sailor.kython.interpreter.pyobject
```
Licence header

Fuyukai commented on src/main/kotlin/interpreter/pyobject/PyBytes.kt:
```diff
@@ -0,0 +1,25 @@
+        return this.wrapped.contentEquals(other.wrapped)
+    }
+
+    override fun hashCode(): Int {
```
`override fun hashCode(): Int = wrapped.contentHashCode()`

Fuyukai commented on src/main/kotlin/interpreter/pyobject/types/PyBytesType.kt:
```diff
@@ -0,0 +1,46 @@
+        try {
+            PyInt(self.wrapped[index.wrappedInt.toInt()].toLong())
+        } catch (e: IndexOutOfBoundsException) {
+            Exceptions.INDEX_ERROR("index out of range").throwKy()
```
Change this to `Index ${index.wrappedInt} out of range`

Martmists-GH replied:
Shouldn't we aim to keep these consistent with cpython's errors?

Fuyukai replied:
I've already decided that no, we shouldn't, because cpython errors are often unhelpful and if you do direct checks on these your code will break if cpy ever gets around to better errors.

Fuyukai commented:
Make sure to pass this through ktlint before the final squash,

Martmists-GH committed:
remove whitespace

Martmists-GH committed:
add bytes to builtins

Martmists-GH committed:
Remove unused import
